programs/gummyroll/src/state/merkle_roll.rs

```diff
@@ -258,82 +326,114 @@ impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> MerkleRoll<MAX_DEPTH,
             } else {
-                updated_leaf = self.change_logs[j as usize].get_leaf();
+                updated_leaf = self.change_logs[changelog_buffer_index as usize].get_leaf();
             }
             // If use_full_buffer is true, this loop will do 1 full pass of the change logs
-            if use_full_buffer && j == self.active_index {
+            if use_full_buffer && changelog_buffer_index == self.active_index {
                 break;
             }
         }
         sol_log_compute_units();
-        let valid_root = recompute(updated_leaf, proof, index) == self.get_change_log().root;
-        if updated_leaf != leaf || index > self.rightmost_proof.index {
+        updated_leaf == leaf
+    }
+
+    /// Fast-forwards submitted proof to be valid for the root at `self.current_index`
+    ///
+    /// Updates proof & updates root & stores
+    ///
+    /// Takes in `j`, which is the root index that this proof was last valid for
+    fn update_and_apply_proof(
+        &mut self,
+        current_root: Node,
+        leaf: Node,
+        new_leaf: Node,
+        proof: &mut [Node; MAX_DEPTH],
+        leaf_index: u32,
+        append_on_conflict: bool,
+        use_full_buffer: bool,
+    ) -> Option<Node> {
+        let mask: usize = MAX_BUFFER_SIZE - 1;
+        let changelog_buffer_index: u64;
+        if use_full_buffer {
+            changelog_buffer_index = self.active_index.wrapping_sub(self.buffer_size) & mask as u64;
+        } else {
+            match self.find_root_in_changelog(current_root) {
+                Some(matching_changelog_index) => {
+                    changelog_buffer_index = matching_changelog_index;
+                }
+                None => {
+                    return None;
+                }
+            }
+        }
+
+        msg!(
+            "Fast-forwarding proof, starting index {}",
+            changelog_buffer_index
+        );
+
+        let valid_fast_forward = self.fast_forward_proof(
+            new_leaf,
+            proof,
+            leaf_index,
+            changelog_buffer_index,
+            use_full_buffer,
+        );
+
+        let valid_root = recompute(leaf, proof, leaf_index) == self.get_change_log().root;
+        if !valid_fast_forward || leaf_index > self.rightmost_proof.index {
             // If the supplied root was not found in the queue, the instruction should fail if the leaf index changes
-            if !use_full_buffer && valid_root && leaf == EMPTY && append_on_conflict {
+            // NOTE: previously we checked if the FF'd proof with the value of the overwritten leaf
+            //      could be hashed to match the current root
+            //      However, this was removed for simplicity because it did not add to the security model
+            //      of insert_or_append functionality.
+            if !use_full_buffer && leaf == EMPTY && append_on_conflict {
```

Comment:
@jarry-xiao I removed `valid_root` check here because if the leaf was updated since the proof was passed in, then the proof should be considered invalid.

Lmk if u disagree. I'm open to redoing this PR that adds these instructions in a less-invasive way.

Reply:
Agreed that this is only for `insert_or_append` instruction.

But `insert_or_append` is still rekt even with `valid_root` check. 
I just `insert_or_append` to any leaf that's in the changelog buffer. Fast forwarding will autcomplete proof & I will always be able to mint.

Am I wrong here?

Reply:
not if you don't have a valid proof!